mirascope/integrations/middleware_factory.py

```diff
@@ -1,154 +1,169 @@
-)
-_F = TypeVar("_F", bound=Callable[..., Any])
-_DecoratorType = Callable[[_F], _F]
+SyncFunc = Callable[
+    _P, _BaseCallResponseT | _BaseStreamT | _BaseModelT | _BaseStructuredStreamT
+]
+AsyncFunc = Callable[
+    _P,
+    Awaitable[_BaseCallResponseT | _BaseStreamT | _BaseModelT | _BaseStructuredStreamT],
+]
 _T = TypeVar("_T")
 
 
 @contextmanager
-def default_context_manager():
+def default_context_manager(
+    fn: SyncFunc | AsyncFunc,
+):
     yield None
 
 
-def middleware(
-    fn: Callable[_P, _R] | Callable[_P, Awaitable[_R]],
+def middleware_decorator(
+    fn: SyncFunc | AsyncFunc,
     *,
     custom_context_manager: Callable[
-        [], Generator[_T, Any, None]
+        [SyncFunc | AsyncFunc], AbstractContextManager[_T]
     ] = default_context_manager,
-    custom_decorator: _DecoratorType | None = None,
+    custom_decorator: SyncFunc | None = None,
     handle_call_response: Callable[[BaseCallResponse, _T | None], None] | None = None,
     handle_call_response_async: Callable[[BaseCallResponse, _T | None], Awaitable[None]]
     | None = None,
     handle_stream: Callable[[BaseStream, _T | None], None] | None = None,
     handle_stream_async: Callable[[BaseStream, _T | None], Awaitable[None]]
     | None = None,
-    handle_base_model: Callable[[BaseModel, _T | None], None] | None = None,
-    handle_base_model_async: Callable[[BaseModel, _T | None], Awaitable[None]]
+    handle_base_model: Callable[[BaseModel | BaseStructuredStream, _T | None], None]
+    | None = None,
+    handle_base_model_async: Callable[
+        [BaseModel | BaseStructuredStream, _T | None], Awaitable[None]
+    ]
     | None = None,
-) -> Callable[_P, _R] | Callable[_P, Awaitable[_R]]:
-    """Wraps a Mirascope function with middleware."""
+) -> SyncFunc | AsyncFunc:
+    is_async = inspect.iscoroutinefunction(fn)
 
-    def inner(*args: _P.args, **kwargs: _P.kwargs) -> _R:
+    @wraps(fn)
+    def inner(*args: _P.args, **kwargs: _P.kwargs) -> SyncFunc:
         result = fn(*args, **kwargs)
         if isinstance(result, BaseCallResponse) and handle_call_response is not None:
-            with custom_context_manager() as context:
+            with custom_context_manager(fn) as context:
                 if context is None:
-                    handle_call_response(result)
+                    handle_call_response(result, None)
                 else:
                     handle_call_response(result, context)
-        elif isinstance(result, BaseStream) and handle_stream is not None:
+        elif isinstance(result, BaseStream):
             original_iter = result.__iter__
 
-            def new_iter(self):
+            def new_stream_iter(self):
                 # Create the context manager when user iterates over the stream
-                with custom_context_manager() as context:
+                with custom_context_manager(fn) as context:
                     for chunk, tool in original_iter():
```

Comment:
does this work? I'm wondering if you'll need to call the iter outside the function and then iterate through the result here instead of calling it here

Easy way to check would be to see if streaming has cost but doesn't when using logfire?

Reply:
streaming does have cost, per my tests so this should work